text/3729-sized-hierarchy.md

```diff
@@ -0,0 +1,1546 @@
+      parameters a less restrictive bound which is not backwards compatible.
+[^6]: If it was desirable to add a new trait to this RFC's proposed hierarchy
+      then there are four possibilities:
+
+      - Before `Sized`
+        - i.e. `NewSized: Sized: ValueSized: Pointee`
+        - Adding bounds on a new trait before `Sized` is not
+          backwards-compatible as described above in [Implementing
+          `Sized`][implementing-sized].
+      - Between `Sized` and `ValueSized`
+        - i.e. `Sized: NewSized: ValueSized: Pointee`
+        - Adding bounds on a new trait between `Sized` and `ValueSized` is
+          backwards compatible for some functions (except for on a trait's
+          associated type). An existing function with a `Sized` bound could
+          be relaxed to the new trait without breaking existing code.
+          Existing functions with a `ValueSized` or `Pointee` bound could not.
+      - Between `ValueSized` and `Pointee`
+        - i.e. `Sized: ValueSized: NewSized: Pointee`
+        - Adding bounds on a new trait between `ValueSized` and `Pointee` is
+          backwards compatible for some functions (except for on a trait's
+          associated type). An existing function with a `Sized` or `ValueSized`
+          bound could be relaxed to the new trait without breaking existing code.
+          Existing functions with a `Pointee` bound could not.
+      - After `Pointee`
+        - i.e. `Sized: ValueSized: Pointee: NewSized`
+        - For the same reasons as the traits proposed in this RFC, adding bounds
+          on a new trait after `Pointee` is backwards compatible (except for on a
+          trait's associated type). Any existing function will have stricter
+          bounds than the new trait and so could be relaxed to the new trait
+          without breaking existing code.
+
+## `Sized` bounds
+[sized-bounds]: #sized-bounds
+
+`?Sized` would be made syntactic sugar for a `const ValueSized` bound. A
+`const ValueSized` bound is equivalent to a `?Sized` bound as all values in Rust
+today whose types do not implement `Sized` are valid arguments to
+[`std::mem::size_of_val`][api_size_of_val] and as such have a size which can be
+computed given a value and knowledge of the type and target platform, and
+therefore will implement `const ValueSized`. As there are currently no
+extern types or other types which would not implement `const ValueSized`,
+every type in Rust today which would satisfy a `?Sized` bound would satisfy
+a `const ValueSized` bound.
+
+**Edition change:** In the current edition,`?Sized` will be syntatic sugar for
+a `const ValueSized` bound. As the `?Trait` syntax is currently accepted for any trait
+but ignored for every trait except `Sized`, `?ValueSized` and `?Pointee` bounds would
+be ignored. In the next edition, any uses of `?Sized` syntax will be rewritten to
+a `const ValueSized` bound. Any other uses of the `?Trait` syntax will be removed as
```

Comment:
I don't think this is possible, there is a future possibility in this RFC that very briefly touches on how we could relax associated types, but I don't think adding implicit `T:Assoc: const ValueSized` bounds would work, because something like this compiles today..

```
trait Foo {
    type Bar;
}

fn qux<T: Foo>() -> usize {
    std::mem::size_of::<<T as Foo>::Bar>()
}
```

..and wouldn't if we added that bound.

Reply:
The idea is that, if we relaxed the bound, it'd work like this:

```rust
trait Tr {
    type Ty: ?Sized default Sized;
    //       ~~~~~~~        ~~~~~
    //  The relaxed bound   |
    //                      ^ The backward compatible default for
    //                        use-site bounds.
}
```

Where at the use site then, if you wrote this...

```rust
fn f<T: Tr> {}
```

...it would be treated according to that `default` as though you had written:

```rust
fn f<T: Tr<Ty: Sized>> {}
//        ~~~~~~~~~~~
//      Added implicitly.
```

This works because loosening the bound doesn't break implementors.  The only reason it breaks callers is because it changes what they can assume by default without explicitly bounding the associated type.  But we could conceivably separate what callers can assume by default from the minimum that implementors are allowed to implement.

Then, conceivably, over an edition, we could change the default.  Use sites in older edition code would still get the old default.  During the edition migration, we would elaborate the code according to that default, which would work in both editions and preserve the semantics.

I asked around about this; unfortunately there are apparently some reasons it could be difficult to implement.